external/loaders/tests/test__config.py

```diff
@@ -0,0 +1,262 @@
+import pytest
+import unittest.mock
+import loaders
+import contextlib
+import xarray as xr
+import tempfile
+import os
+import yaml
+import dataclasses
+
+
+@contextlib.contextmanager
+def registration_context(registration_dict):
+    """
+    A context manager that provides a clean slate for registering functions,
+    and restores the registration state when exiting.
+    """
+    original_functions = {}
+    original_functions.update(registration_dict)
+    registration_dict.clear()
+    yield
+    registration_dict.clear()
+    registration_dict.update(original_functions)
+
+
+@contextlib.contextmanager
+def mapper_context():
+    with registration_context(loaders._config.mapper_functions):
+        mock_mapper = {"key": xr.Dataset()}
+        mock_mapper_function = unittest.mock.MagicMock(return_value=mock_mapper)
+        mock_mapper_function.__name__ = "mock_mapper_function"
+        loaders._config.mapper_functions.register(mock_mapper_function)
+        yield mock_mapper_function
```

Comment:
The layered contexts for adding in mocked functions still seems overly-complex but maybe it is the best answer.  I wonder if we could hide some of this by using pytest's `monkeypatch` to temporarily set the module registers.

e.g.:

```
class MockRegister(loader._config.Register()
@pytest.fixture
def mock_mapper():
    return {"key": xr.Dataset()}

@pytest.fixture
def mock_batches():
    return [xr.Dataset()]

def get_magic_mock(name, return_val):
    mock = unittest.moc.MagicMock(return_value=return_val)
    mock.__name__ = name
    return mock

@pytest.fixture
def mock_mapper_func(mock_mapper, monkeypatch):

    # anything using this fixture will have the temporarily cleared register
    monkeypatch.setattr(loaders._config, "mapper_functions", loaders._config.Register())
    mocked_func = get_magic_mock("mock_mapper_function", mock_mapper)

    return loaders._config.mapper_functions.register(mocked_func)

def test_load_mapper(mock_mapper, mock_mapper_func)
    data_path = "test/data/path"
    mapper_kwargs = {"arg1": "value1", "arg2": 2}
    config = loaders._config.MapperConfig(
        data_path=data_path,
        mapper_function="mock_mapper_function",
        mapper_kwargs=mapper_kwargs,
    )
    result = config.load_mapper()
    assert result is mock_mapper
    mock_mapper_function.assert_called_once_with(data_path, **mapper_kwargs)
```

Something like this would let you load the expected return value and the function with adjusted registry all just using fixtures instead of context adjustments.

Reply:
The only issue with true monkeypatching (e.g. with monkeypatch) is that these objects are generally not used as `import module; module.object` but rather as `from module import object`, to reduce line length. Monkeypatching would not replace occurences of these registry dictionaries that have been imported in this manner, which could lead to bugs (including bugs from future changes). Since I already had things working via a clear-and-restore-dictionary context manager, I opted to keep it that way instead.

> Something like this would let you load the expected return value and the function with adjusted registry all just using fixtures instead of context adjustments.

This isn't strictly true - monkeypatch uses a context manager to achieve the same effect, just under the hood. If the issue is the `with` statements, I could turn these into fixtures instead if that's desirable. I just thought it's clearer that it's manipulating things external to the tests when it's a context manager.

Reply:
> This isn't strictly true - monkeypatch uses a context manager to achieve the same effect, just under the hood. If the issue is the with statements, I could turn these into fixtures instead if that's desirable. I just thought it's clearer that it's manipulating things external to the tests when it's a context manager.

Sorry, you're right about what I wrote at the end.  What I had in mind was that it might be better to hide the complexity of the layered context management innards from future us.  But I can see the argument for making it clear that the test is modifying external state while running.


> Monkeypatching would not replace occurrences of these registry dictionaries that have been imported in this mannerMonkeypatching would not replace occurrences of these registry dictionaries that have been imported in this manner

The `setitem` monkeypatch for could be used instead of replacing the module object, but that gives less lee-way to other manual adjustments to the register.

The reason I looked into this was mostly that the code looked quite complex and still has a number of repeats.  When that happens it may make future us more hesitant to touch/work on it.  Moving contexts into fixtures would reduce the nesting a bit and make for a standard way of getting testing info to multiple places, but then obscure the contexts switching.  Although, obscuring could also be a point of confusion for further test development. 

Maybe just the context naming could be more illustrative like `clean_batches_register` and then a `_get_magic_mock` helper would be good.